pkg/cmd/template/cmd_overlays_test.go

```diff
@@ -146,6 +146,195 @@ yamlfunc: yamlfunc`)
 	}
 }
 
+func TestDocumentOverlaysMapItems(t *testing.T) {
+	yamlTplData := []byte(`
+---
+clients:
+  client1:
+    secret: foo
+  client2:
+    secret: bar
+  client3:
+    needsSecret: true
+`)
+
+	yamlOverlayTplData := []byte(`
+#@ load("@ytt:overlay", "overlay")
+#@overlay/match by=overlay.all
+---
+clients:
+  #@overlay/match by=overlay.all,expects="1+"
+  _:
+    #@overlay/match missing_ok=True
+    secret: ree
+    #@overlay/match missing_ok=True
+    #@overlay/remove
+    needsSecret: true
+`)
+
+	expectedYAMLTplData := `clients:
+  client1:
+    secret: ree
+  client2:
+    secret: ree
+  client3:
+    secret: ree
+`
+
+	filesToProcess := files.NewSortedFiles([]*files.File{
+		files.MustNewFileFromSource(files.NewBytesSource("tpl.yml", yamlTplData)),
+		files.MustNewFileFromSource(files.NewBytesSource("overlay.yml", yamlOverlayTplData)),
+	})
+
+	ui := cmdcore.NewPlainUI(false)
+	opts := cmdtpl.NewOptions()
+
+	out := opts.RunWithFiles(cmdtpl.TemplateInput{Files: filesToProcess}, ui)
+	if out.Err != nil {
+		t.Fatalf("Expected RunWithFiles to succeed, but was error: %s", out.Err)
+	}
+
+	if len(out.Files) != 1 {
+		t.Fatalf("Expected number of output files to be 1, but was %d", len(out.Files))
+	}
+
+	file := out.Files[0]
+
+	if file.RelativePath() != "tpl.yml" {
+		t.Fatalf("Expected output file to be tpl.yml, but was %#v", file.RelativePath())
+	}
+
+	if string(file.Bytes()) != expectedYAMLTplData {
+		t.Fatalf("Expected output file to have specific data, but was: >>>%s<<<", file.Bytes())
+	}
+}
+
+func TestDocumentOverlaysMapItemsBySubset(t *testing.T) {
+	yamlTplData := []byte(`
+---
+clients:
+  client1:
+    secret: foo
+  client2:
+    secret: bar
+  client3:
+    needsSecret: true
+`)
+
+	yamlOverlayTplData := []byte(`
+#@ load("@ytt:overlay", "overlay")
+#@overlay/match by=overlay.all
+---
+clients:
+  #@overlay/match by=overlay.subset({"client3": {"needsSecret": True}})
```

Comment:
should this be `by=overlay.subset({"needsSecret": True})` because we are matching on value part of mapitem (ie client3 is key part of mapitem)?

Reply:
Yes.... yes, it should. ... and makes me suspicious of `MapItem` implementation of `pullOutMapValue()` below (see [comments](https://github.com/k14s/ytt/pull/93/files/14e537d3f63ad991af1517e1c19b3eab5188f4a3#r382657540), below).

Digging into this...